Knapsack: Proposal for new exercise (#1484)
exercism/problem-specifications: exercises/knapsack/description.md

> Given a set of items, each with a weight, w and a value, v, determine the number of each item to include in a collection so that the total weight is less than or equal to a given limit and the total value is as large as possible

Reviewer: this sentence sort of implies the general knapsack problem to me because "number of each item" hasn't told me that it's restricted to 1 and 0. It seems to me like that is an important aspect of the problem, and that if we can express it here, we should do so to help students understand what they have to plan for.

Author: Agreed, I have tweaked the description and hopefully made it more interesting.

> Given a set of items, each with a weight, w and a value, v, determine the number of each item to include in a collection so that the total weight is less than or equal to a given limit and the total value is as large as possible

Reviewer: this sentence says "determine the number of each item", but that is not consistent with the "Answer is 90" line below. If the answer is 90, we determined the maximum value, not the number of items. The two sentences should be made consistent with each other (and also with canonical-data.json).

Author: Yup changed as well!

Knapsack Max weight: W = 10 

> Total items: N = 4

Reviewer: Total items (N) doesn't seem necessary to specify as an input; it can be inferred from the sizes of `v[]` and `w[]`.

If I think about how this problem would be implemented in a variety of languages:

Some languages will benefit from explicitly specifying N.
Some will not.
I predict that languages that benefit from explicitly specifying N are in the minority. This says how much benefit we get.

If this repository specifies N in canonical-data.json, the reviewers of this file will then be responsible for making sure it matches the lengths of `v[]` and `w[]`, with the associated costs.

An automated reviewer of course would be more able to bear this cost. Note that https://github.com/exercism/problem-specifications/issues/460 indicates a lack of interest in setting up an automated reviewer.

Author: Yup understood. Modified this as well.

Values of items: v[] = {10, 40, 30, 50}
Weight of items: w[] = {5, 4, 6, 3}

> Answer is 90, you can take item 1 and item 4 (total weight is 7)

Reviewer 2: item 2 and 4, I presume?

Author: Sry, yes

In this exercise, lets try to solve a classic problem.

Bob is a thief. After months of carefuly planning, he finally manages to crack the security systems of a high-class apartment. 

In front of him are many items, each with a value (v) and weight (w). Bob of course wants to maximize the total value he can get; he would gladly take all of the items if he could. However, to his horror, he realized that the knapsack he carried with him can only hold so much weight(W).

Given a knapsack with a specific capacity (W), help Bob determine the maximum value he can get from the items in the house. Note, Bob can only take only one of each item. 

> All values given will be strictly positive integers.Items will be represented as a list of pairs, [wi, vi] where the first element `wi` is the weight of the ith item and `vi` is the value for that item.

Reviewer 3: ```suggestion
All values given will be strictly positive integers.  Items will be represented as a list of pairs, [wi, vi] where the first element `wi` is the weight of the ith item and `vi` is the value for that item.
```
Need a space (or two) after period.

In this exercise, lets try to solve a classic problem.

> Bob is a thief. After months of planning, he finally managed to crack the security systems of a high-class apartment. In front of him are many items, each with a value (V). Bob of course wants to maximise the total value he can get; he would of course take all the items if he could. However, to his horrer, he realised that the knapsack he carried with him can only hold a weight W.

Author: You are right. I have updated the description.

Given a knapsack of weight W, can you help Bob determine the maximum value he can get from the items in the house? Note, you can take only one of each item.

For example:

Item Values: [10, 40, 30, 50]

> Item Weight: [5, 4, 6, 4]

Author: Yup updated the description and made it a list of pairs.

Knapsack Limit: 10

For the above, item 1 has weight 5 and value 10, items 2 has weight 4 and value 40, and so on...

Bob should take item 2 and 4 to maximise his value which in this case is 90. He cannot get more than 90 as his knapsack has a limit of 10

> Follow-Up:

Author: Yup understood. I have removed it for now.

In this exercise, lets try to solve a classic problem.

Bob is a thief. After months of carefuly planning, he finally manages to crack the security systems of a high-class apartment. 

In front of him are many items, each with a value (v) and weight (w). Bob of course wants to 
maximize the total value he can get; he would gladly take all of the items if he could. 
However, to his horror, he realized that the knapsack he carried with him can only hold so much weight(W).

> Given a knapsack with a specific capacity (W), can you help Bob determine the maximum value he can get from the items in the house? Note, Bob can only take only one of each item. 

Reviewer 3: ```suggestion
Given a knapsack with a specific capacity (W), help Bob determine the maximum value he can get from the items in the house. Note, Bob can only take only one of each item. 
```
The question is OK, My personal preference is to have a statement that states what needs to be done.  One only needs to answer `Yes` to the question.

Reviewer 3: If you turn the sentence around you can eliminate the comma.
```
Help Bob determine the maximum value he can get from the items in the house given a knapsack with a specific capacity (W).
```

Author: Yup I think a statement is good as well!

All values given will be strictly positive integers.Items will be represented as a list of pairs, 

> [wi, vi] where the first element (wi) is the weight of the ith item and vi is the value for 

Reviewer 3: ```suggestion
[wi, vi] where the first element (wi) is the weight of the ith item and (vi) is the value for 
```

that item.

For example:

Items: [[5, 10], [4, 40], [6, 30], [4, 50]]
Knapsack Limit: 10

> 

Reviewer 3: minor detail, remove extra blank line.

Author: Done!

In this exercise, lets try to solve a classic problem.

Bob is a thief. After months of planning, he 
finally manages to crack the security systems of a 
high-class apartment. In front of him are many 

> items, each with a value (v) and weight (w). Bob of

Reviewer 3: ```suggestion
items, each with a value (v) and weight (w). Bob of,
```

> course wants to maximise the total value he can 

Reviewer 3: ```suggestion
course, wants to maximise the total value he can 
```

> course wants to maximise the total value he can 

Reviewer 3: ```suggestion
course wants to maximize the total value he can 
```

Reviewer 4: Is there are British/American language policy on Exercism?

Reviewer 5: Not that I have bloody heard, mate!  But I think we allow them.

Reviewer 3: I didn't realize `maximise` was British.  Most words are are easier to spot IE. harbor / harbour.  OK how do I force Chrome to use American and British dictionaries?.....  OK done.  Now Chrome knows the Queen's English.  😃   Carry on.

get; he would of course take all the items if he

> could. However, to his horrer, he realised that the 

Reviewer 3: ```suggestion
could. However, to his horror, he realized that the 
```

Reviewer 4: It seems that I've suggested this correction once already.

> knapsack he carried with him can only hold a weight W.

Reviewer 3: ```suggestion
knapsack he carried with him can only hold so much weight (W).
```

> Given a knapsack of weight W, can you help Bob 

Reviewer 3: ```suggestion
Given a knapsack of weight limit (W), can you help Bob 
```

determine the maximum value he can get from the 
items in the house? Note, you can take only one of 
each item. 

All values given will be strictly positive integers.

> Items will be represented as a list of pairs (wi, vi) where the first element (wi) is the weight of the ith item and vi is the value for that item.

Reviewer 3: ```suggestion
Items will be represented as a list of pairs [wi, vi] where the first element (wi) is the weight of the ith item and (vi) is the value for that item.
```

Reviewer 4: @rpottsoh: I proposed to change this into

    a list of pairs, `wi` and `vi`, ...

to avoid concrete syntax for pairs.

For example:

> Items: [(5, 10), (4, 40), (6, 30), (4, 50)]

Reviewer 3: ```suggestion
Items: [[5, 10], [4, 40], [6, 30], [4, 50]]
```

In this exercise, lets try to solve a classic problem.

Bob is a thief. After months of careful planning, he finally manages to
crack the security systems of a high-class apartment.

In front of him are many items, each with a value (v) and weight (w). Bob,
of course, wants to maximize the total value he can get; he would gladly
take all of the items if he could. However, to his horror, he realizes that
the knapsack he carries with him can only hold so much weight (W).

Given a knapsack with a specific carrying capacity (W), help Bob determine
the maximum value he can get from the items in the house. Note that Bob can

> only take only one of each item.

Reviewer 4: ```suggestion
only take one of each item.
```

One "only" too many.

Author: Updated!

> In this exercise, lets try to solve a classic problem.

Reviewer 2: lets => let's

Reviewer 6: For convenience:

```suggestion
In this exercise, let's try to solve a classic problem.
```

Author: Changed

Bob is a thief. After months of careful planning, he finally manages to
crack the security systems of a high-class apartment.

In front of him are many items, each with a value (v) and weight (w). Bob,
of course, wants to maximize the total value he can get; he would gladly
take all of the items if he could. However, to his horror, he realizes that
the knapsack he carries with him can only hold so much weight (W).

Given a knapsack with a specific carrying capacity (W), help Bob determine
the maximum value he can get from the items in the house. Note that Bob can

> only take one of each item.

Reviewer 5: ~~only~~ take only one of each item.

Because if Bob can only take one of each item it does not allow the option of taking only one of each item.

Author: Yes good catch! I have updated this.

In this exercise, lets try to solve a classic problem.

Bob is a thief. After months of careful planning, he finally manages to
crack the security systems of a high-class apartment.

In front of him are many items, each with a value (v) and weight (w). Bob,
of course, wants to maximize the total value he can get; he would gladly
take all of the items if he could. However, to his horror, he realizes that
the knapsack he carries with him can only hold so much weight (W).

Given a knapsack with a specific carrying capacity (W), help Bob determine
the maximum value he can get from the items in the house. Note that Bob can
only take only one of each item.

> All values given will be strictly positive. Items will be represented as a

Reviewer 4: Oh, I also changd "positive integers" to "positive", since this exercise may be implemented for languages that don't differentiate between integers and floating-point numbers.

Author: Yup no problem 👍
